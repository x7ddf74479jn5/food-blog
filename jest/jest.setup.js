// Optional: configure or set up a testing framework before each test.
// If you delete this file, remove `setupFilesAfterEnv` from `jest.config.js`

// Used for __tests__/testing-library.js
// Learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";

window.matchMedia = jest.fn().mockImplementation((query) => {
  return {
    addListener: jest.fn(),
    matches: false,
    media: query,
    onchange: null,
    removeListener: jest.fn(),
  };
});
