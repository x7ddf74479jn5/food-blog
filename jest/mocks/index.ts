export { mockNextImage } from "./image";
export type { NextRouter } from "./router";
export { mockRouter, RouterContext, withMockedRouter } from "./router";
export { setupWindow } from "./window";
