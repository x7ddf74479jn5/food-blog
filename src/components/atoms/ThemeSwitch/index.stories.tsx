import type { ComponentMeta, ComponentStory } from "@storybook/react";

import ThemeSwitch from ".";

export default {
  title: "Atoms/ThemeSwitch",
  component: ThemeSwitch,
} as ComponentMeta<typeof ThemeSwitch>;

const Template: ComponentStory<typeof ThemeSwitch> = (args) => <ThemeSwitch {...args} />;

export const Default = Template.bind({});
