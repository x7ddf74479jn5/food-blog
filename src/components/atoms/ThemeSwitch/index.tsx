import { useTheme } from "next-themes";
import { useEffect, useState } from "react";
import { IconContext } from "react-icons";
import { FaMoon, FaSun } from "react-icons/fa";
import Switch from "react-switch";

const ThemeSwitch: React.FC = () => {
  const { theme, setTheme } = useTheme();

  const isDark = theme === "dark" ? true : false;

  const [isChecked, setChecked] = useState(isDark);
  const [isMounted, setMounted] = useState(false);

  const handleChange = (nextChecked: boolean) => {
    setChecked(nextChecked);
  };

  // When mounted on client, now we can show the UI
  useEffect(() => setMounted(true), []);

  useEffect(() => {
    setTheme(isChecked ? "dark" : "light");
  }, [isChecked, setTheme]);

  if (!isMounted) return null;

  return (
    <Switch
      onChange={handleChange}
      checked={isChecked}
      aria-label="switch between day and night themes"
      offColor="#555"
      onHandleColor="#eee"
      handleDiameter={20}
      uncheckedIcon={
        <div className="flex h-full items-center justify-center">
          <IconContext.Provider
            value={{
              color: "gold",
              size: "80%",
            }}
          >
            <FaSun />
          </IconContext.Provider>
        </div>
      }
      checkedIcon={
        <div className="flex h-full items-center justify-center">
          <IconContext.Provider
            value={{
              color: "yellow",
              size: "80%",
            }}
          >
            <FaMoon />
          </IconContext.Provider>
        </div>
      }
      height={24}
      width={48}
    />
  );
};

export default ThemeSwitch;
