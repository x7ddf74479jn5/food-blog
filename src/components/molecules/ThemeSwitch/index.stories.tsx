import type { ComponentMeta, ComponentStory } from "@storybook/react";
import { ThemeProvider } from "next-themes";

import { ThemeSwitch } from ".";

export default {
  title: "molecules/ThemeSwitch",
  component: ThemeSwitch,
} as ComponentMeta<typeof ThemeSwitch>;

const Template: ComponentStory<typeof ThemeSwitch> = (args) => (
  <ThemeProvider enableSystem>
    <ThemeSwitch {...args} />
  </ThemeProvider>
);

export const Default = Template.bind({});
