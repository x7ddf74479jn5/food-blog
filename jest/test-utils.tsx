import type Queries from "@testing-library/dom/types/queries";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { RequestHandler } from "msw";
import { setupServer } from "msw/node";
import { SWRConfig } from "swr";

import { SearchProvider } from "@/contexts/search/SearchContext";
import { ViewportProvider } from "@/contexts/viewport/ViewportContext";

import { withMockLegacyRouter, withMockRouter } from "./mocks";

export const Providers: React.ComponentType<{ children?: React.ReactNode }> = ({ children }) => {
  return withMockRouter(
    withMockLegacyRouter(
      <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
        <SearchProvider>
          <ViewportProvider>{children}</ViewportProvider>
        </SearchProvider>
      </SWRConfig>
    )
  );
};

const customRender = (ui: React.ReactElement, options = {}): RenderResult<typeof Queries, HTMLElement> => {
  return render(ui, { wrapper: Providers, ...options });
};

// re-export everything
export * from "./mocks";
export * from "@testing-library/react";

// override render method
export { render as originalRender, customRender as render };

export const reTestCase = {
  anyImage: expect.stringMatching(/^(data:image\/gif)|\.(png|webp|jpeg|jpg|svg)$/),
  anyWord: expect.stringMatching(/\w+/),
};

type ErrorComponentProps = {
  message?: string;
};
export const ErrorComponent: React.FC<ErrorComponentProps> = ({ message }) => {
  throw new Error(message);
};

export const setupMockServer = (...handlers: RequestHandler[]) => {
  const server = setupServer(...handlers);
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());
  return server;
};
