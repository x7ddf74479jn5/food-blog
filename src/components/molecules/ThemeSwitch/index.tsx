"use client";

import { useTheme } from "next-themes";
import { useCallback, useEffect } from "react";
import { IconContext } from "react-icons";
import { FaMoon, FaSun } from "react-icons/fa";
import Switch from "react-switch";

import { Skelton } from "@/components/atoms/Skelton";
import { useMount } from "@/hooks/useMount";

const useSwitch = (isMounted: boolean) => {
  const { resolvedTheme, setTheme, theme } = useTheme();
  const isSystem = theme === "system";
  const isChecked = resolvedTheme === "dark";

  const handleToggleSwitch = useCallback(() => {
    setTheme(isChecked ? "light" : "dark");
  }, [isChecked, setTheme]);

  useEffect(() => {
    let isCleanup = false;
    // 初回はtheme="system"なのでスキップ
    if (!isCleanup && !isSystem) {
      setTheme(isChecked ? "dark" : "light");
    }

    return () => {
      isCleanup = true;
    };
  }, [isChecked, isMounted, isSystem, setTheme]);

  return { handleToggleSwitch, isChecked, isMounted };
};

export const ThemeSwitch: React.FC = () => {
  const isMounted = useMount();
  const { handleToggleSwitch, isChecked } = useSwitch(isMounted);

  if (!isMounted) return <Skelton className="h-6 w-12" />;

  return (
    <Switch
      onChange={handleToggleSwitch}
      checked={isChecked}
      aria-label="switch between day and night themes"
      offColor="#555"
      onHandleColor="#eee"
      handleDiameter={20}
      uncheckedIcon={
        <div className="flex h-full items-center justify-center">
          <IconContext.Provider
            value={{
              color: "gold",
              size: "80%",
            }}
          >
            <FaSun />
          </IconContext.Provider>
        </div>
      }
      checkedIcon={
        <div className="flex h-full items-center justify-center">
          <IconContext.Provider
            value={{
              color: "yellow",
              size: "80%",
            }}
          >
            <FaMoon />
          </IconContext.Provider>
        </div>
      }
      height={24}
      width={48}
    />
  );
};
