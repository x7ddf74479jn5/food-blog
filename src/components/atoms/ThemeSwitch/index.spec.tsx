import { render } from "jest/test-utils";

import ThemeSwitch from ".";

describe("components/atoms/ThemeSwitch", () => {
  it("snapshot", () => {
    const { asFragment } = render(<ThemeSwitch />);
    expect(asFragment()).toMatchSnapshot();
  });
});
