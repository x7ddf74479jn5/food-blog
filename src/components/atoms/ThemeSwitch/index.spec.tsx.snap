// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/atoms/ThemeSwitch snapshot 1`] = `
<DocumentFragment>
  <div
    style="position: relative; display: inline-block; text-align: left; opacity: 1; direction: ltr; border-radius: 12px; transition: opacity 0.25s; user-select: none;"
  >
    <div
      class="react-switch-bg"
      style="height: 24px; width: 48px; margin: 0px; position: relative; background: rgb(85, 85, 85); border-radius: 12px; cursor: pointer; transition: background 0.25s;"
    >
      <div
        style="height: 24px; width: 27px; position: relative; opacity: 0; pointer-events: none; transition: opacity 0.25s;"
      >
        <div
          class="flex justify-center items-center h-full"
        >
          <svg
            fill="currentColor"
            height="80%"
            stroke="currentColor"
            stroke-width="0"
            style="color: yellow;"
            viewBox="0 0 512 512"
            width="80%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M283.211 512c78.962 0 151.079-35.925 198.857-94.792 7.068-8.708-.639-21.43-11.562-19.35-124.203 23.654-238.262-71.576-238.262-196.954 0-72.222 38.662-138.635 101.498-174.394 9.686-5.512 7.25-20.197-3.756-22.23A258.156 258.156 0 0 0 283.211 0c-141.309 0-256 114.511-256 256 0 141.309 114.511 256 256 256z"
            />
          </svg>
        </div>
      </div>
      <div
        style="height: 24px; width: 27px; position: absolute; opacity: 1; right: 0px; top: 0px; pointer-events: none; transition: opacity 0.25s;"
      >
        <div
          class="flex justify-center items-center h-full"
        >
          <svg
            fill="currentColor"
            height="80%"
            stroke="currentColor"
            stroke-width="0"
            style="color: gold;"
            viewBox="0 0 512 512"
            width="80%"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
            />
          </svg>
        </div>
      </div>
    </div>
    <div
      class="react-switch-handle"
      style="height: 20px; width: 20px; background: rgb(255, 255, 255); display: inline-block; cursor: pointer; border-radius: 50%; position: absolute; transform: translateX(2px); top: 2px; outline: 0; border: 0px; transition: background-color 0.25s, transform 0.25s, box-shadow 0.15s;"
    />
    <input
      aria-checked="false"
      aria-label="switch between day and night themes"
      role="switch"
      style="border: 0px; height: 1px; margin: -1px; overflow: hidden; padding: 0px; position: absolute; width: 1px;"
      type="checkbox"
    />
  </div>
</DocumentFragment>
`;
